design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+
+#### Flow
+
+The flow for Milestone 1 is similar to the user experience from Phase 1. For each step, we will go into more detail below.
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret ***(Milestone 1)***
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role for the Service Accounts with **get**/**update** and also **list** privileges on K8s Secrets resources ***(Milestone 1)***
+
+4. Create RoleBinding and bind the Service Accounts to the role from previous step
+
+5. Create/Deploy Secrets Providers as Jobs in deployment manifest(s) ***(Milestone 1)***
+   - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s) ***(Milestone 1)***
+
+#### Customer Experience
+
+In order to preserve restrictions on Conjur secrets, a customer can decide to either:
+
+1. Deploy 1 Secrets Provider per namespace, giving the Secrets Provider access to all secrets
+2. Deploy 1 Secrets Provider per app / (group of apps), giving the Secrets Provider only access to the secrets they have been given access to in Conjur. Should the user require further separation of duties on Conjur secrets, they will need to deploy another Secrets Provider
+
+The Secrets Provider host will resemble the following:
+
+```yaml
+- !policy
+  id: secrets
+  body:
+  - !host
+    id: secrets-provider-1
+    annotations:
+      authn-k8s/namespace: prod-namespace
+      authn-k8s/service-account: prod-sa
+      authn-k8s/authenticator-container-name: cyberark-secrets-provider
+
+  - !layer 
+
+  - !permit
+    resources: !variable super_secret
+    role: !layer secrets-accessors
+    privileges: [ read, execute ]
+ 
+- !grant
+  role: !layer secrets-accessors
+  members:
+    - !host conjur/authn-k8s/secrets-provider-1
+```
+
+The Secrets Provider manifest will be defined as the following:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: secrets-provider-job
+spec:
+  template:
+    spec:
+      serviceAccountName: prod-sa
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider-1
+
+# Rest of environment variables for Pod to run 
+```
+
+We recommend that the admin will define "Role" instead of "ClusterRole" to follow the principle of least privilege on resources.
+
+#### Code changes
+
+1. Support Job application identity 
+2. Enhance Batch retrieval
+
+*Support Job application identity*
+
+To authenticate to Conjur, the Secrets Provider application identity (or Resource Restrictions) will be the characteristics of the Pod that it is running in. Currently, we do not have a "Job" application identity granularity so to allow the Secrets Provider the ability authenticate to Conjur, the customer will need to choose between the Namespace or Service Account granularities. If there is a requirement to support a "Job" granularity this can be done by adding code to the authenticator in Conjur.
```

Comment:
so what's the decision? will we implement this as part of the effort or no?

Reply:
i would also add the location in which the change will occur. in this case it is in the Conjur server.